### PR TITLE
[KYUUBI #4623][K8S] KubernetesApplicationOperation uses Informer instead of list

### LINKED ISSUE #4623: [Improvement][K8S] kubernetesApplicationOperation Using Informer instead of list
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

KubernetesApplicationOperation makes too many requests to the kubernetes api server (in list form) when a large number of tasks are submitted at the same time, causing some stress.

### How should we improve?

https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/SharedInformerExample.java

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by zwangsheng**
### _Why are the changes needed?_

Close #4623 

To reduce the pressure on the Api Server (which use the kubernetes client polls with label to find the `spark driver pod` when multiple Rest Application are running at the same time), use informer, the kubernetes-recommended method of maintaining the application state.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request
- [x] Run CI

### CODE REVIEW

**Comment by zwangsheng** [comment]
Fix On Kubernetes IT Test, ready for review @pan3793

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 1]
```diff
@@ -127,22 +137,68 @@ class KubernetesApplicationOperation extends ApplicationOperation with Logging {
...
   }
 
-  override def stop(): Unit = {
-    if (kubernetesClient != null) {
-      try {
-        kubernetesClient.close()
-      } catch {
-        case e: Exception => error(e.getMessage)
+  class DriverPodEventHandler extends ResourceEventHandler[Pod] {
+    private def filter(pod: Pod): Boolean = {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:157]: the method name is confusing

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 2]
```diff
@@ -127,22 +137,68 @@ class KubernetesApplicationOperation extends ApplicationOperation with Logging {
...
-    val size = podList.size()
-    if (size != 1) {
-      warn(s"Get Tag: ${tag} Driver Pod In Kubernetes size: ${size}, we expect 1")
+  override def stop(): Unit = {
+    try {
+      if (kubernetesClient != null) {
+        kubernetesClient.close()
+      }
+      if (driverInformer != null) {
+        driverInformer.stop()
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:146]: is the stop order right?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala (hunk 1 above)
Comment (zwangsheng) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:157]: Renamed to `SparkEnginePodEventHandler`

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 3]
```diff
@@ -17,30 +17,56 @@
...
 
+  import io.fabric8.kubernetes.api.model.Pod
+
   @volatile
   private var kubernetesClient: KubernetesClient = _
-
+  private var driverInformer: SharedIndexInformer[Pod] = _
   private var submitTimeout: Long = _
 
+  private val appInfoStore: ConcurrentHashMap[String, ApplicationInfo] =
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:42]: when to delete the appInfo?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 4]
```diff
@@ -127,22 +137,68 @@ class KubernetesApplicationOperation extends ApplicationOperation with Logging {
...
+        updateState(pod)
+      }
+    }
+
+    override def onUpdate(oldPod: Pod, newPod: Pod): Unit = {
+      if (isSparkEnginePod(newPod)) {
+        updateState(newPod)
+        toApplicationState(newPod.getStatus.getPhase) match {
+          case FINISHED | FAILED | UNKNOWN =>
+            markDeleted(newPod)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:191]: it's confusing, delete what? Pod or Application?

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:191]: And from the K8s docs description about Pod state, "Unknown" sounds like a transient error

> Unknown 
>
> For some reason the state of the Pod could not be obtained. This phase typically occurs due to an error in communicating with the node where the Pod should be running.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -1171,6 +1171,24 @@ object KyuubiConf {
       .booleanConf
       .createWithDefault(false)
 
+  val KUBERNETES_INFORMER_PERIOD: ConfigEntry[Long] =
+    buildConf("kyuubi.kubernetes.informer.period")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1175]:
```suggestion
    buildConf("kyuubi.kubernetes.informer.resyncPeriod")
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -1171,6 +1171,24 @@ object KyuubiConf {
...
+      .doc("Kubernetes Informer poll driver pod period." +
+        "Set too small can lead to, stress on Kubernetes Api Server; " +
+        "Set too lager can lead to, app info can't updated in time ")
+      .version("1.8.0")
+      .timeConf
+      .checkValue(_ > 0, "must be positive number")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1181]: From the JavaDoc of `io.fabric8.kubernetes.client.dsl.Informable#runnableInformer`

> the resync period or 0 for no resync

Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1181]:
```suggestion
      .checkValue(_ >= 0, "Invalid resync period provided, It should be a non-negative value")
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala (hunk 4 above)
Comment (zwangsheng) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:191]: Using `ApplicationState#isTerminated` instead.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -1171,6 +1171,14 @@ object KyuubiConf {
       .booleanConf
       .createWithDefault(false)
 
+  val KUBERNETES_INFORMER_CACHE_PERIOD: ConfigEntry[Long] =
+    buildConf("kyuubi.kubernetes.informer.cache.period")
+      .doc("The time kyuubi server hold app info after the app ends.")
+      .version("1.8.0")
+      .timeConf
+      .checkValue(_ > 0, "must be positive number")
+      .createWithDefault(Duration.ofSeconds(60).toMillis)
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1180]:
```suggestion
    val KUBERNETES_TERMINATED_APPLICATION_RETAIN_PERIOD: ConfigEntry[Long] =
    buildConf("kyuubi.kubernetes.terminatedApplicationRetainPeriod")
      .doc("The period for which the Kyuubi server retains application information after " +
        "the application terminates.")
      .version("1.8.0")
      .timeConf
      .checkValue(_ > 0, "must be positive number")
      .createWithDefault(Duration.ofMinutes(5).toMillis)
```

**Review by pan3793 — APPROVED** [review]

File: integration-tests/kyuubi-kubernetes-it/src/test/scala/org/apache/kyuubi/kubernetes/test/spark/SparkOnKubernetesTestsSuite.scala [hunk 8]
```diff
@@ -208,19 +208,20 @@ class KyuubiOperationKubernetesClusterClusterModeSuite
       batchRequest.getConf.asScala.toMap,
       batchRequest)
 
-    val session = sessionManager.getSession(sessionHandle).asInstanceOf[KyuubiBatchSessionImpl]
-    val batchJobSubmissionOp = session.batchJobSubmissionOp
+    sessionManager.getSession(sessionHandle).asInstanceOf[KyuubiBatchSessionImpl]
```
Comment (pan3793) [inline integration-tests/kyuubi-kubernetes-it/src/test/scala/org/apache/kyuubi/kubernetes/test/spark/SparkOnKubernetesTestsSuite.scala:211]: why change it?

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master/1.7

**Pull request closed without merging**